packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -4,94 +4,253 @@
-        ]
-        // We can only update cache for this product list if we have the ID
         const listId = response.id
-        if (!listId) return {invalidate}
         return {
-            invalidate,
-            update: [{queryKey: getCustomerProductList.queryKey({...parameters, listId})}]
+            update: [
+                {
+                    queryKey: getCustomerProductLists.queryKey(parameters),
+                    updater: createUpdateFunction((result: CustomerProductListResult) => {
+                        // Push new address onto the end of addresses list.
+                        // Add new list to front of the lists.
+                        result.data.unshift(response)
+                        result.limit++
+                        result.total++
+
+                        return result
+                    })
+                },
+                {
+                    queryKey: getCustomerProductList.queryKey({...parameters, listId})
+                }
+            ]
         }
     },
     createCustomerProductListItem(customerId, {parameters}, response) {
         // We always invalidate, because even without an ID we assume that something has changed
-        // TODO: Rather than invalidate, can we selectively update?
-        const invalidate: CacheUpdate['invalidate'] = [
-            {queryKey: getCustomerProductList.queryKey(parameters)},
-            {queryKey: getCustomerProductLists.queryKey(parameters)}
-        ]
-        // We can only update cache for this product list item if we have the ID
+        // QUESTION: Why would we get a response that doesn't have an ID?
         const itemId = response.id
-        if (!itemId) return {invalidate}
+
         return {
-            invalidate,
-            update: [{queryKey: getCustomerProductListItem.queryKey({...parameters, itemId})}]
+            update: [
+                {
+                    queryKey: getCustomerProductListItem.queryKey({...parameters, itemId})
+                },
+                {
+                    queryKey: getCustomerProductList.queryKey(parameters),
+                    updater: createUpdateFunction((list: CustomerProductList) => {
+                        return {
+                            ...list,
+                            customerProductListItems: [
+                                ...(list?.customerProductListItems || []),
+                                response
+                            ]
+                        }
+                    })
+                },
+                {
+                    queryKey: getCustomerProductLists.queryKey(parameters),
+                    updater: createUpdateFunction((result: CustomerProductListResult) => {
+                        // Find the list that we want to add the item to.
+                        const list = result.data.find(({id}) => id === parameters.listId)
+
+                        // Push the new item onto the end of the list.
```

Comment:
```suggestion
                        // Push the new item onto the end of the list.
```
Feels like this comment is in the wrong spot! 😉